evidence/reactor.go

```diff
@@ -131,90 +133,50 @@ func (evR *Reactor) broadcastEvidenceRoutine(peer p2p.Peer) {
 			}
 		}
 
-		ev := next.Value.(types.Evidence)
-		msg, retry := evR.checkSendEvidenceMessage(peer, ev)
-		if msg != nil {
-			success := peer.Send(EvidenceChannel, cdc.MustMarshalBinaryBare(msg))
-			retry = !success
-		}
-
-		if retry {
-			time.Sleep(peerCatchupSleepIntervalMS * time.Millisecond)
+		peerState, ok := peer.Get(types.PeerStateKey).(PeerState)
+		if !ok {
+			// Peer does not have a state yet. We set it in the consensus reactor, but
+			// when we add peer in Switch, the order we call reactors#AddPeer is
+			// different every time due to us using a map. Sometimes other reactors
+			// will be initialized before the consensus reactor. We should wait a few
+			// milliseconds and retry.
+			time.Sleep(broadcastEvidenceIntervalS * time.Second)
 			continue
 		}
 
-		afterCh := time.After(time.Second * broadcastEvidenceIntervalS)
-		select {
-		case <-afterCh:
-			// start from the beginning every tick.
-			// TODO: only do this if we're at the end of the list!
-			next = nil
-		case <-next.NextWaitChan():
-			// see the start of the for loop for nil check
-			next = next.Next()
-		case <-peer.Quit():
-			return
-		case <-evR.Quit():
-			return
+		var evMsg ListMessage
+		for next != nil && len(evMsg.Evidence) < messageSize {
+			ev := next.Value.(types.Evidence)
+			// check to see if peer is behind, if so jump to next piece of evidence
+			if ev.Height() <= peerState.GetHeight() {
+				// Update should ensure that all evidence in the list is not expired
+				// hence we don't need to check it
```

Comment:
Would you agree with this @melekes or do you think we should check that the evidence hasn't expired

Reply:
I agree